metrics/network/network-metrics-iperf3.sh

```diff
@@ -0,0 +1,567 @@
+			"Units"  : "$total_bidirectional_client_bandwidth_units"
+		},
+		"server to client": {
+ 			"Result" : $total_bidirectional_server_bandwidth,
+			"Units"  : "$total_bidirectional_server_bandwidth_units"
+		}
+	}
+EOF
+)"
+
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Results"
+	metrics_json_save
+	clean_env
+}
+
+# This function checks/verify if the iperf3 server
+# is ready/up for requests.
+function check_iperf3_server() {
+	local retry=6
+	local period=0.5
+	local test_cmd="iperf3 -c "$server_address" -t 1"
+
+	# check tools dependencies
+	local cmds=("netstat")
+	check_cmds "${cmds[@]}"
+
+	while [ 1 ]; do
+		if ! bash -c "$test_cmd" > /dev/null 2>&1; then
+			echo "waiting for server..."
+			count=$((count++))
+			sleep $period
+		else
+			echo "iperf3 server is up!"
+			break;
+		fi
+
+		if [ "$count" -le "$retry" ]; then
+			die "iperf3 server init fails"
+		fi
+	done
+
+	# Check listening port
+	lport="$(netstat -atun | grep "$port" | grep "LISTEN")"
+	if [ -z "$lport" ]; then
+		die "port is not listening"
+	fi
+}
+
+# This function parses the output of iperf3 execution
+function parse_iperf3_bwd() {
+	local TEST_NAME="$1"
+	local result="$2"
+
+	if [ -z "$result" ]; then
+		die "no result output"
+	fi
+
+	metrics_json_init
+	save_config
+
+	# Filter receiver/sender results
+	rx_res=$(echo "$result" | grep "receiver" | awk -F "]" '{print $2}')
+	tx_res=$(echo "$result" | grep "sender" | awk -F "]" '{print $2}')
+
+	metrics_json_start_array
+
+ 	# Getting results
+	rx_bwd=$(echo "$rx_res" | awk '{print $5}')
+	tx_bwd=$(echo "$tx_res" | awk '{print $5}')
+	rx_uts=$(echo "$rx_res" | awk '{print $6}')
+	tx_uts=$(echo "$tx_res" | awk '{print $6}')
+
+	local json="$(cat << EOF
+	{
+ 		"receiver": {
+			"Result" : $rx_bwd,
+			"Units"  : "$rx_uts"
+		},
+		"sender": {
+			"Result" : $tx_bwd,
+			"Units"  : "$tx_uts"
+		}
+	}
+EOF
+)"
+
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Results"
+	metrics_json_save
+}
+
+# This function parses the output of iperf3 UDP execution, and
+# saves the receiver successful datagram value in the results.
+function parse_iperf3_pps() {
+	local TEST_NAME="$1"
+	local result="$2"
+
+	if [ -z "$result" ]; then
+		die "no result output"
+	fi
+
+	metrics_json_init
+	save_config
+
+	metrics_json_start_array
+
+	# Extract results
+	local lost=$(echo "$result" | jq '.end.sum.lost_packets')
+ 	local total=$(echo "$result" | jq '.end.sum.packets')
+	local notlost=$((total-lost))
+	local pps=$((notlost/transmit_timeout))
+
+	local json="$(cat << EOF
+	{
+		"receiver": {
+			"Result" : $pps,
+			"Units"  : "PPS"
+		},
+		"lost":     {
+			"Result" : $lost,
+			"Units"  : "PPS"
+		},
+		"total":    {
+			"Result" : $total,
+			"Units"  : "PPS"
+		},
+		"not lost":  {
+			"Result" : $notlost,
+			"Units"  : "PPS"
+		}
+	}
+EOF
+)"
+
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Results"
+	metrics_json_save
```

Comment:
I still think we can add a json save here of the value of `$result`, as I think that is a fully qualified json item - it might have a little JSON wrap around it that we may need to trim in order to embed it. But, seeing as we have a full JSON formatted detailed result for this test, I think we should store it in the results file rather than throw away any details. I grabbed some output. It is quite large, so here is a trimmed example:

```json
{
        "start":        {
                "connected":    [{
                                "socket":       4,
                                "local_host":   "172.17.0.1",
                                "local_port":   59172,
                                "remote_host":  "172.17.0.2",
                                "remote_port":  5201
                        }],
                "version":      "iperf 3.0.11",
                "system_info":  "Linux bignuc 4.4.0-104-generic #127-Ubuntu SMP Mon Dec 11 12:16:42 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux\n",
                "timestamp":    {
                        "time": "Tue, 24 Jul 2018 12:44:50 GMT",
                        "timesecs":     1532436290
                },
                "connecting_to":        {
                        "host": "172.17.0.2",
                        "port": 5201
                },
                "cookie":       "bignuc.1532436290.341903.6630386652f",
                "test_start":   {
                        "protocol":     "UDP",
                        "num_streams":  1,
                        "blksize":      64,
                        "omit": 0,
                        "duration":     30,
                        "bytes":        0,
                        "blocks":       0,
                        "reverse":      1
                }
        },
        "intervals":    [{
                        "streams":      [{
                                        "socket":       4,
                                        "start":        0,
                                        "end":  1.00001,
                                        "seconds":      1.00001,
                                        "bytes":        12268992,
                                        "bits_per_second":      9.8151e+07,
                                        "jitter_ms":    0.00722195,
                                        "lost_packets": 651136,
                                        "packets":      842839,
                                        "lost_percent": 77.2551,
                                        "omitted":      false
                                }],
                        "sum":  {
                                "start":        0,
                                "end":  1.00001,
                                "seconds":      1.00001,
                                "bytes":        12268992,
                                "bits_per_second":      9.8151e+07,
                                "jitter_ms":    0.00722195,
                                "lost_packets": 651136,
                                "packets":      842839,
                                "lost_percent": 77.2551,
                                "omitted":      false
                        }
                }, {
...                }],
        "end":  {
                "streams":      [{
                                "udp":  {
                                        "socket":       4,
                                        "start":        0,
                                        "end":  30,
                                        "seconds":      30,
                                        "bytes":        1601305600,
                                        "bits_per_second":      4.27015e+08,
                                        "jitter_ms":    0.0010719,
                                        "lost_packets": 18732833,
                                        "packets":      25020345,
                                        "lost_percent": 74.8704
                                }
                        }],
                "sum":  {
                        "start":        0,
                        "end":  30,
                        "seconds":      30,
                        "bytes":        1601305600,
                        "bits_per_second":      4.27015e+08,
                        "jitter_ms":    0.0010719,
                        "lost_packets": 18732833,
                        "packets":      25020345,
                        "lost_percent": 74.8704
                },
                "cpu_utilization_percent":      {
                        "host_total":   57.289,
                        "host_user":    11.7294,
                        "host_system":  45.5612,
                        "remote_total": 99.7418,
                        "remote_user":  19.7103,
                        "remote_system":        80.0314
                }
        }
}
```

I'd like us to store that in the JSON results in a new section (maybe called `raw` for example).
Of course, ideally I'd probably now convert all the `iperf3` tests to output json results natively, which we could store - but that's a moderate rewrite. the PPS tests are easy, as they already generate JSON which they then parse.

Reply:
@grahamwhaley all the changes were applied, I implemented the function `raw_iperf3_pps_results` to save all the rest of the information